Block method for Slate expressions

thomasgibson opened the pull request:
I've added a `.block()` method which pulls out blocks of a Slate expression. I use the argument indices to specify a particular block.

wence- commented on firedrake/slate/slate.py:
```diff
@@ -296,6 +305,107 @@ def _key(self):
+            raise ValueError("Can only index block (mixed) tensors")
+
+        fs_extent = len(tensor.arguments()[0].function_space())
+        if any(i not in range(0, fs_extent) for i in idx):
```
This seems wrong.  I think you want:

```python
for arg, i in zip(tensor.arguments(), idx):
    if not 0 <= i < len(arg.function_space()):
        raise ValueError(...)
```

wence- commented on firedrake/slate/slate.py:
```diff
@@ -296,6 +305,107 @@ def _key(self):
+        )
+
+        if not tensor.is_mixed:
+            raise ValueError("Can only index block (mixed) tensors")
```
I think it makes sense to be able to index non-mixed (assuming that the indices are all 0).  This makes writing code outside much more natural (you don't have to special-case outside).  

You code do:

```python
def __new__(cls, tensor, idx):
    ...
    if not tensor.is_mixed:
        assert all(i == 0 for i in idx)
        return tensor
    return super().__new__(cls)
```

wence- commented on firedrake/slate/slate.py:
```diff
@@ -296,6 +305,107 @@ def _key(self):
+
+        assert len(idx) == tensor.rank, (
+            "Number of indices must match the tensor rank."
+        )
```
This case also catches indexing a "scalar" with more than zero indices.  So again, I think it makes sense to allow `IndexedTensor(scalar, ())`

wence- commented on firedrake/slate/slate.py:
```diff
@@ -332,15 +442,12 @@ def __init__(self, form):
-        for it in integrals:
-            map_integrand_dags(mapper, it)
+        # Remove any negative restrictions and replace with zero
+        _form = map_integrand_dags(RemoveNegativeRestrictions(), form)
```
You don't need `form` after this call, so just reassign `form = map_integrand_dags(..., form)`

wence- commented on firedrake/slate/slac/compiler.py:
```diff
@@ -477,24 +476,51 @@ def metaphrase_slate_to_cpp(expr, temps, prec=None):
+        tensor, = expr.operands
+        idx = expr._idx
+        if expr.rank == 1:
+            # Vector segment
```
WAT?!  How does eigen not treat vectors just like matrices :P.

thomasgibson replied:
Ah, I just remembered that all of these `Eigen::MatrixBase` objects (which includes vectors as well) have a `block` method. So I'm pretty sure I can squash this into one stream of code for all cases.

wence- commented:
So this doesn't solve the (difficult) problem of deciding which sub-blocks actually need to be assembled to provide a particular indexed block.  But perhaps that make sense because you're usually needing all of them and then pulling them apart?

thomasgibson commented:
Correct. In all cases I've looked at, all blocks are used in computations. I think it would be nice to detect which bit needs assembly and only compute what is necessary, but I haven't gotten around to implementing it just yet. I'll update soon addressing your comments!

thomasgibson commented:
I've pushed some changes. Let me know if this looks alright

wence- commented:
My only real concern is that going forward, we might want to index slate expressions by quadrature-point or basis function indices.  So I think it would be good to think about what we want the indexing space of slate expressions to be.  Should it be the same as UFL forms?  This isn't quite that, since here the index space is of arguments, whereas for forms it is the `ufl_shape` of the expression.

I'm not opposed to the idea of indexing tensors, I just want to have a discussion about what type of indexing we want, and whether or not this approach precludes indexing over different spaces in the future.

Perhaps we need to having different index types? (That feels ugly though).

thomasgibson commented:
Once build issues are resolved, I think this is ready. As we've discussed, I decided to remove indexing for now but provide the same feature as a `.block()` method. This takes argument indices just like the `ExtractSubBlock.split()` method in `formmanipulation.py`.

This works for extracting blocks over one or multiple test/trial spaces. I've tested in this my own implementations of static condensation for CG problems and fully formed hybridized mixed/DG problems. I added some tests in Firedrake, but it's currently not possible to test assembly of individual blocks spanning multiple spaces. 'Cause, you know, mixed mats/dats.

wence- commented on firedrake/slate/slate.py:
```diff
@@ -140,6 +141,42 @@ def inv(self):
+            assert self.rank == 2, "Rank %d not supported." % self.rank
+            arg_indices = (to_tuple(test_idx), to_tuple(trial_idx))
+
+        return Block(self, arg_indices)
```
I'd prefer to be liberal in what one receives here.  I think you should be allowed to `block` a 0-tensor, by passing `tensor.block(()) => tensor`.  This way you don't have to write special-case code outside.

Similarly, you could (like `ExtractSubBlock`) just accept any number of indices, but only handle the ones you care about.

wence- replied:
But if you're going to do all this type-checking, best to do it in `Block`, since then you get it when either the user calls `Tensor.block`, or else when they call `Block` directly.

thomasgibson replied:
Okay I'll update this!

wence- commented on firedrake/slate/slate.py:
```diff
@@ -296,6 +333,116 @@ def _key(self):
+        """Constructor for the Block class."""
+        assert len(arg_indices) == tensor.rank, (
+            "Number of indices must match the tensor rank."
+        )
```
I continue to be sad at the extra line breaks.  Just write `assert foo, "bar"`.  We don't have a strict 80 character line limit.

thomasgibson replied:
Sorry... it's a habit.

wence- commented on firedrake/slate/slate.py:
```diff
@@ -296,6 +333,116 @@ def _key(self):
+
+        super(Block, self).__init__()
+        self.operands = (tensor,)
+        self._blocks = dict(zip((0, 1), arg_indices))
```
I realise a better way to write this is `dict(enumerate(arg_indices))`

wence- commented on firedrake/slate/slate.py:
```diff
@@ -296,6 +333,116 @@ def _key(self):
+        spaces determined by the indices.
+        """
+        from firedrake.functionspace import (FunctionSpace,
+                                             MixedFunctionSpace)
```
No need for a line break here.

wence- commented on tests/slate/test_assemble_tensors.py:
```diff
@@ -170,6 +170,76 @@ def test_mixed_argument_tensor(mesh):
+    assert np.allclose(A10.M.values, A10ref.M.values, rtol=1e-14)
+    assert np.allclose(A11.M.values, A11ref.M.values, rtol=1e-14)
+    assert np.allclose(A20.M.values, A20ref.M.values, rtol=1e-14)
+    assert np.allclose(A22.M.values, A22ref.M.values, rtol=1e-14)
```
Can you test that taking a `block` resulting in a mixed matrix, and then taking a block of that (into a non-mixed matrix) and then assembling that works too?

wence- replied:
I note in passing that if you made the API consistent between form-splitting and tensor blocking, this test could be much reduced in length.

thomasgibson commented on firedrake/slate/slate.py:
```diff
@@ -140,6 +141,42 @@ def inv(self):
     def T(self):
         return Transpose(self)
 
+    def block(self, test_idx, trial_idx=None):
```
I chose `block` because that seemed to be the logical choice for name here. I want "a block of a matrix over argument indices (...)" I'm not opposed to renaming it, but there was reason and thought into this.

I think `split` could work. So this probably means this method should return several `Block` objects corresponding to the function arguments provided?

thomasgibson commented:
OK I've updated this branch. The API here should be more similar to form splitting. Test suite now includes more robust block testing.
